src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs

```diff
@@ -147,5 +147,51 @@ public IDataView FilterByKeyColumnFraction(IDataView input, string columnName, d
             return new RangeFilter(Environment, input, columnName, lowerBound, upperBound, false);
         }
+
+        /// <summary>
+        /// Shuffle the rows of <paramref name="input"/>.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="Shuffle"/> will shuffle the rows of any input <see cref="IDataView"/> using a streaming approach.
+        /// In order to not load the entire dataset in memory, a pool of <paramref name="shufflePoolSize"/> rows will be used
+        /// to randomly select rows to output. The pool is constructed from the first <paramref name="shufflePoolSize"/> rows
+        /// in <paramref name="input"/>. Rows will then be randomly yielded from the pool and replaced with the next row from <paramref name="input"/>
+        /// until all the rows have been yielded, resulting in a new <see cref="IDataView"/> of the same size as <paramref name="input"/>
+        /// but with the rows in a randomized order.
+        /// If the <see cref="IDataView.CanShuffle"/> property of <paramref name="input"/> is true, then it will also be read into the
+        /// pool in a random order, offering two sources of randomness.
+        /// </remarks>
+        /// <param name="input">The input data.</param>
+        /// <param name="seed">The random seed. If unspecified, the random state will be instead derived from the <see cref="MLContext"/>.</param>
+        /// <param name="shufflePoolSize">The number of rows to hold in the pool. Setting this to 1 will turn off pool shuffling and
+        /// <see cref="Shuffle"/> will only perform a shuffle by reading <paramref name="input"/> in a random order.</param>
+        /// <param name="shuffleSource">If false, the transform will not attempt to read <paramref name="input"/> in a random order and only use
+        /// pooling to shuffle. This parameter has no effect if the <see cref="IDataView.CanShuffle"/> property of <paramref name="input"/> is false.
+        /// </param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[Shuffle](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Shuffle.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public IDataView Shuffle(IDataView input,
+            uint? seed = null,
+            int shufflePoolSize = RowShufflingTransformer.Defaults.PoolRows,
+            bool shuffleSource = !RowShufflingTransformer.Defaults.PoolOnly)
+        {
+            Environment.CheckValue(input, nameof(input));
+            Environment.CheckUserArg(shufflePoolSize > 0, nameof(shufflePoolSize), "Pool size must be positive");
+
+            var options = new RowShufflingTransformer.Options
+            {
+                PoolRows = shufflePoolSize,
+                PoolOnly = !shuffleSource,
+                ForceShuffle = true,
+                ForceShuffleSeed = (int)seed
```

Comment:
>(int)seed [](start = 35, length = 9)

If this is left `null`, which is the default value, this cast will throw an exception. I am pretty sure that you meant to do something like `seed ?? ...get random seed from Environment...` or somesuch.

Reply:
Also changed the parameter to BootstrapSample to be a `int?` and cast to `uint?`.

---
In reply to: [254346714](https://github.com/dotnet/machinelearning/pull/2427#discussion_r254346714) [](ancestors = 254346714)